### DISCUSSION TITLE
getting qualified type names of parameters (and defined classes) from pmd's java ast

### DISCUSSION THREAD

**Discussion by trktby** [pmd/pmd#4136] [Q&A] [answered]
i'm trying to use some of pmds features programmatically among other libraries to extract some java source code metrics. so i need to get the qualified name of classes and method parameters in order to build a consistent way of addressing methods/classes across different libraries.

for the sake of illustration, a very artificial method declared as 
```
public class VariousArgumentTypes {
  class InnerType {
  }

  int mixedArguments(int i, int[] is, String s, InnerType[] its) {
    return i;
  }
}
```
should compute to a signature ` VariousArgumentTypes.mixedArguments(int, int[], java.lang.String, VariousArgumentTypes$InnerType[])`

while trying to get the types of the parameters, according to the documentation on [formal parameters](https://www.javadoc.io/static/net.sourceforge.pmd/pmd-java/6.49.0/net/sourceforge/pmd/lang/java/ast/ASTFormalParameter.html#getType()) i should be able to use getType() with a caveat that it *may* return null. i don't know what *may* means specifically but so far for me, it has always returned null.
```    
ASTFormalParameters params = methodDeclaration.getFormalParameters();
for (ASTFormalParameter param : params) {
  System.out.println(param.getType());
}
```
prints null for every source code parameter type i've tried, which includes all primitives, arrays, varargs, java.lang objects like String, objects needing imports like Set<String> and custom classes declared as subclasses of the one i'm building the ast for. so calling getName() of course throws a NullPointerException.
the same behaviour also occurs when using `param.getVariableDeclaratorId().getType()` as well as `param.getVariableDeclaratorId().getNameDeclaration()`, `param.getTypeDefinition()` and `param.getTypeNode().getType()` .

the closest i could get was through `param.getTypeNode().getTypeImage()` which however isn't specific enough for my needs.

similarly, for classes (accessed as variables of type `ASTAnyTypeDeclaration`) [the documentation](https://www.javadoc.io/static/net.sourceforge.pmd/pmd-java/6.49.0/net/sourceforge/pmd/lang/java/ast/ASTAnyTypeDeclaration.html) states i should use `getBinaryName()`. but this also triggers NullPointerExceptions on every source class i tried.

i'm running this from a maven project with the following pmd dependency:
```
<dependency>
  <groupId>net.sourceforge.pmd</groupId>
  <artifactId>pmd-java</artifactId>
  <version>6.49.0</version>
</dependency>
```
i hope someone here can help me with this issue.

(i've also asked the same question on stack [overflow](https://stackoverflow.com/questions/73772974/getting-qualified-type-names-of-parameters-and-defined-classes-from-pmds-java))

**Comment by jsotuyod** [comment, marked as answer]
> according to the documentation on [formal parameters](https://www.javadoc.io/static/net.sourceforge.pmd/pmd-java/6.49.0/net/sourceforge/pmd/lang/java/ast/ASTFormalParameter.html#getType()) i should be able to use getType() with a caveat that it may return null. i don't know what may means specifically but so far for me, it has always returned null.

Usually, a type may be null if not available in the auxclasspath provided during analysis.

However, since in your case, even java types are not being resolved, it seems like you may have only parsed the Java files without actually running the type resolution stage (hard to tell as we don't know how you are using PMD, and this is certainly a non-standard use case). I suggest you take a look at [SourceCodeProcessor.processSource](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/SourceCodeProcessor.java#L195-L210) as for the different processing stages that enrich the AST with symbol resolution, type resolution, etc.

**Reply by adangel** [reply]
Also have a look at https://pmd.github.io/latest/pmd_userdocs_tools_java_api.html on how to call PMD programmatically.
The auxclasspath needs to be provided via the configuration (`configuration.prependAuxClasspath(...)`). 
This all of course describes the standard use case: executing rules.

**Reply by trktby** [reply]
thank you!
i did indeed not run through the type resolution stage.
after doing so the same way SourceCodeProcessor does it, things resolve as expected for now.
